designs/2021-stricter-rule-test-validation/README.md

```diff
@@ -0,0 +1,279 @@
+- Repo: eslint/eslint
+- Start Date: 2021-11-30
+- RFC PR: <https://github.com/eslint/rfcs/pull/84>
+- Authors: [bmish](https://github.com/bmish)
+
+# Stricter validation of rule tests
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+We will add additional assertions to the rule test runner to address several potential points of poor test coverage, especially related to [rule suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions).
+
+Assertions added to invalid test case error objects:
+
+- Must contain `suggestions` if the test case produces suggestions
```

Comment:
@mdjermanovic thanks for calling this out. I think we should enforce this even for test cases that don't have error objects. Why? To be consistent with test cases that produce autofixes. Test cases that produce autofixes cannot be written with the above shorthands because the autofix is required to be tested.

Reply:
> Test cases that produce autofixes cannot be written with the above shorthands because the autofix is required to be tested.

`output` is a top-level property so it can be used in combination with `errors: <number>` and other shorthands for `errors`. We have such tests in the ESLint codebase:

https://github.com/eslint/eslint/blob/18f5e05bce10503186989d81ca484abb185a2c9d/tests/lib/rules/comma-dangle.js#L1796-L1824

Those are mostly tests where we run multiple rules to check whether the combined autofix breaks the code, so  `output` is the main concern, and `errors` would have to specify errors for fixture rules so we left out the details.

Reply:
Ah, my bad, thank you for the correction. So my recommendation is still that when suggestions are produced, we require them to be tested, meaning the error shorthand can't be used then. Does that sound good?

Reply:
I think it makes sense for `errors: ["message text", /message regex/]`, as these are just shorthands for `errors: [{ message: "message text" }, { message: /message regex/ }]`.

But I'm not sure about `errors: 2`. That looks like an opt-out from testing any details about the errors.

Reply:
The `errors: 2` was the first option because we didn’t really know the best way to test things at the start. I think this is probably a terrible way to test rules now. :)

Reply:
The `errors: <number>` and `suggestions: <number>` generally aren't the best way to write tests but I'm fine to allow these both in all situations. Updated.